In [61]:
library(tidyverse)
library(ggplot2)
library(tidyr)
options(scipen=999)
library(readr)


## Import

In [55]:
df_who <- who
print(df_who)

# A tibble: 7,240 x 60
   country  iso2  iso3   year new_sp_m014 new_sp_m1524 new_sp_m2534 new_sp_m3544
   <chr>    <chr> <chr> <dbl>       <dbl>        <dbl>        <dbl>        <dbl>
 1 Afghani~ AF    AFG    1980          NA           NA           NA           NA
 2 Afghani~ AF    AFG    1981          NA           NA           NA           NA
 3 Afghani~ AF    AFG    1982          NA           NA           NA           NA
 4 Afghani~ AF    AFG    1983          NA           NA           NA           NA
 5 Afghani~ AF    AFG    1984          NA           NA           NA           NA
 6 Afghani~ AF    AFG    1985          NA           NA           NA           NA
 7 Afghani~ AF    AFG    1986          NA           NA           NA           NA
 8 Afghani~ AF    AFG    1987          NA           NA           NA           NA
 9 Afghani~ AF    AFG    1988          NA           NA           NA           NA
10 Afghani~ AF    AFG    1989          NA           NA           NA           NA
# i 7

In [59]:
df_pop <- read.csv("Population.csv", skip = 3)
print(paste("Shape:",nrow(df_pop), "x", ncol(df_pop)))
print(df_pop)

[1] "Shape: 266 x 69"
                                            Country.Name Country.Code
1                                                  Aruba          ABW
2                            Africa Eastern and Southern          AFE
3                                            Afghanistan          AFG
4                             Africa Western and Central          AFW
5                                                 Angola          AGO
6                                                Albania          ALB
7                                                Andorra          AND
8                                             Arab World          ARB
9                                   United Arab Emirates          ARE
10                                             Argentina          ARG
11                                               Armenia          ARM
12                                        American Samoa          ASM
13                                   Antigua and Barbuda          AT

#### Shapes
`population` dataset have *266 rows and 69 columns*

`who` dataset have *7240 rows and 60 columns*

#### Summary

In [62]:
summary(df_pop)

 Country.Name       Country.Code       Indicator.Name     Indicator.Code    
 Length:266         Length:266         Length:266         Length:266        
 Class :character   Class :character   Class :character   Class :character  
 Mode  :character   Mode  :character   Mode  :character   Mode  :character  
                                                                            
                                                                            
                                                                            
                                                                            
     X1960                X1961                X1962           
 Min.   :      2715   Min.   :      2970   Min.   :      3264  
 1st Qu.:    515203   1st Qu.:    525523   1st Qu.:    536302  
 Median :   3659633   Median :   3747132   Median :   3831900  
 Mean   : 115447973   Mean   : 117053757   Mean   : 119216045  
 3rd Qu.:  26862930   3rd Qu.:  27613257   3rd Qu.:  28373020  


In [63]:
str(df_pop)

'data.frame':	266 obs. of  69 variables:
 $ Country.Name  : chr  "Aruba" "Africa Eastern and Southern" "Afghanistan" "Africa Western and Central" ...
 $ Country.Code  : chr  "ABW" "AFE" "AFG" "AFW" ...
 $ Indicator.Name: chr  "Population, total" "Population, total" "Population, total" "Population, total" ...
 $ Indicator.Code: chr  "SP.POP.TOTL" "SP.POP.TOTL" "SP.POP.TOTL" "SP.POP.TOTL" ...
 $ X1960         : num  54922 130072080 9035043 97630925 5231654 ...
 $ X1961         : num  55578 133534923 9214083 99706674 5301583 ...
 $ X1962         : num  56320 137171659 9404406 101854756 5354310 ...
 $ X1963         : num  57002 140945536 9604487 104089175 5408320 ...
 $ X1964         : num  57619 144904094 9814318 106388440 5464187 ...
 $ X1965         : num  58190 149033472 10036008 108772632 5521981 ...
 $ X1966         : num  58694 153281203 10266395 111246953 5581386 ...
 $ X1967         : num  58990 157704381 10505959 113795019 5641807 ...
 $ X1968         : num  59069 162329396 10756

In [9]:
print(sapply(df_pop, function(x) sum(is.na(x))))

   country       year population 
         0          0          0 


There doesn't seem to be any anomaly in the data of `population`. However, if we look at the the `who` dataset numerous invalid entry can be found:

In [11]:
print(head(sapply(df_who, function(x) sum(is.na(x))), 10))

     country         iso2         iso3         year  new_sp_m014 new_sp_m1524 
           0           34            0            0         4067         4031 
new_sp_m2534 new_sp_m3544 new_sp_m4554 new_sp_m5564 
        4034         4021         4017         4022 


## Data Transformation
It is more likely that the columns of `who` dataset are values instead of variables. Let's explore this more.

In [20]:
df_who2 <- df_who %>%
    pivot_longer(cols=-c(1:4), names_to="variables", values_to="cases", values_drop_na=TRUE)
head(df_who2, 10)

country     iso2 iso3 year variables    cases
1  Afghanistan AF   AFG  1997 new_sp_m014   0   
2  Afghanistan AF   AFG  1997 new_sp_m1524 10   
3  Afghanistan AF   AFG  1997 new_sp_m2534  6   
4  Afghanistan AF   AFG  1997 new_sp_m3544  3   
5  Afghanistan AF   AFG  1997 new_sp_m4554  5   
6  Afghanistan AF   AFG  1997 new_sp_m5564  2   
7  Afghanistan AF   AFG  1997 new_sp_m65    0   
8  Afghanistan AF   AFG  1997 new_sp_f014   5   
9  Afghanistan AF   AFG  1997 new_sp_f1524 38   
10 Afghanistan AF   AFG  1997 new_sp_f2534 36

In [21]:
unique(df_who2$variables)

[1] "new_sp_m014"  "new_sp_m1524" "new_sp_m2534" "new_sp_m3544" "new_sp_m4554"
 [6] "new_sp_m5564" "new_sp_m65"   "new_sp_f014"  "new_sp_f1524" "new_sp_f2534"
[11] "new_sp_f3544" "new_sp_f4554" "new_sp_f5564" "new_sp_f65"   "new_sn_m014" 
[16] "new_sn_m1524" "new_sn_m2534" "new_sn_m3544" "new_sn_m4554" "new_sn_m5564"
[21] "new_sn_m65"   "new_ep_m014"  "new_ep_m1524" "new_ep_m2534" "new_ep_m3544"
[26] "new_ep_m4554" "new_ep_m5564" "new_ep_m65"   "new_sn_f014"  "newrel_m014" 
[31] "newrel_f014"  "new_sn_f1524" "new_sn_f2534" "new_sn_f3544" "new_sn_f4554"
[36] "new_sn_f5564" "new_sn_f65"   "new_ep_f014"  "new_ep_f1524" "new_ep_f2534"
[41] "new_ep_f3544" "new_ep_f4554" "new_ep_f5564" "new_ep_f65"   "newrel_m1524"
[46] "newrel_m2534" "newrel_m3544" "newrel_m4554" "newrel_m5564" "newrel_m65"  
[51] "newrel_f1524" "newrel_f2534" "newrel_f3544" "newrel_f4554" "newrel_f5564"
[56] "newrel_f65"

##### Data Structure
Looking the variables columns, it appears that there is a clear structure of values. There is however peculiar values by the end that includes the word "newrel." This doesn't make sense because all other values are divided into three. 

Fortunately there exist a data dictionary in [GitHub](https://github.com/GTB-TME/gtbreport2024/blob/main/dynamic/TB_data_dictionary_2022-05-31.csv).
"newrel" as it turns out mean "New and Relapse cases." Separating "new" and "rel" should be fine for this alligns with the structure of the rest of the values under `variables` column.

In [24]:
df_who2$variables <- gsub(pattern = "newrel", replacement = "new_rel", df_who2$variables)
unique(df_who2$variables)

[1] "new_sp_m014"   "new_sp_m1524"  "new_sp_m2534"  "new_sp_m3544" 
 [5] "new_sp_m4554"  "new_sp_m5564"  "new_sp_m65"    "new_sp_f014"  
 [9] "new_sp_f1524"  "new_sp_f2534"  "new_sp_f3544"  "new_sp_f4554" 
[13] "new_sp_f5564"  "new_sp_f65"    "new_sn_m014"   "new_sn_m1524" 
[17] "new_sn_m2534"  "new_sn_m3544"  "new_sn_m4554"  "new_sn_m5564" 
[21] "new_sn_m65"    "new_ep_m014"   "new_ep_m1524"  "new_ep_m2534" 
[25] "new_ep_m3544"  "new_ep_m4554"  "new_ep_m5564"  "new_ep_m65"   
[29] "new_sn_f014"   "new_rel_m014"  "new_rel_f014"  "new_sn_f1524" 
[33] "new_sn_f2534"  "new_sn_f3544"  "new_sn_f4554"  "new_sn_f5564" 
[37] "new_sn_f65"    "new_ep_f014"   "new_ep_f1524"  "new_ep_f2534" 
[41] "new_ep_f3544"  "new_ep_f4554"  "new_ep_f5564"  "new_ep_f65"   
[45] "new_rel_m1524" "new_rel_m2534" "new_rel_m3544" "new_rel_m4554"
[49] "new_rel_m5564" "new_rel_m65"   "new_rel_f1524" "new_rel_f2534"
[53] "new_rel_f3544" "new_rel_f4554" "new_rel_f5564" "new_rel_f65"

Now let us separate the `variables` column into three new columns, `new`, `ep`, and ``

In [28]:
df_who3 <- separate_wider_delim(df_who2, variables, delim = "_", names = c("new", "type", "gender_age"))
head(df_who3, 5)

country     iso2 iso3 year new type gender_age cases
1 Afghanistan AF   AFG  1997 new sp   m014        0   
2 Afghanistan AF   AFG  1997 new sp   m1524      10   
3 Afghanistan AF   AFG  1997 new sp   m2534       6   
4 Afghanistan AF   AFG  1997 new sp   m3544       3   
5 Afghanistan AF   AFG  1997 new sp   m4554       5

In [43]:
df_who_tidy <- df_who2 %>%
    separate_wider_delim(variables, delim = "_", names = c("new", "type", "gender_age")) %>%
    mutate(
        sex=str_sub(gender_age, 1, 1), 
        age_group = str_sub(gender_age, 2, str_length(gender_age))) %>%
    subset(select = -c(gender_age, new))
head(df_who_tidy, 5)

country     iso2 iso3 year type cases sex age_group
1 Afghanistan AF   AFG  1997 sp    0    m   014      
2 Afghanistan AF   AFG  1997 sp   10    m   1524     
3 Afghanistan AF   AFG  1997 sp    6    m   2534     
4 Afghanistan AF   AFG  1997 sp    3    m   3544     
5 Afghanistan AF   AFG  1997 sp    5    m   4554

In [46]:
sort(unique(df_who_tidy$year))

[1] 1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992 1993 1994
[16] 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007 2008 2009
[31] 2010 2011 2012 2013

In [47]:
unique(df_pop$year)

[1] 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007 2008 2009
[16] 2010 2011 2012 2013